src/adaptors/articles.ts

```diff
@@ -0,0 +1,13 @@
+import { Iri, JsonLdObj } from 'jsonld/jsonld-spec';
+
+export interface Articles extends Iterable<JsonLdObj> {
+    count(): Promise<number>;
+
+    get(id: Iri): Promise<JsonLdObj>;
+
+    set(article: JsonLdObj): Promise<void>;
```

Comment:
Should probably call this `add`, since it's not (visibly) assigning the object to something.

Question about if it's clear that it `update`s too...

Reply:
Though there's a question here about the identifier. Currently it's the full IRI/URI, and is using the scheme/host/port at runtime https://github.com/libero/article-store/blob/10f133673ef1abe8cc9597a535b6128adf087394/src/routes/article-list.ts#L10

I've never been entirely happy with this (what happens if it changes?), so maybe it is time to introduce a base URI setting that it used everywhere instead (and is hard/impossible to change). Means if the app doesn't have to care if it's behind a proxy.

Reply:
@stephenwf @tpluscode Any advice?

Reply:
Just so I understand:
* A document list is served with a generated `@id` property based on the origin
* POST-ed documents will have an `@id` in the body of their document
* POST-ed documents will hit an endpoint `/articles/{identifier}`
* Documents will be saved with an `@id` based on [ ... ]
* Documents will be served with an `@id` based on [ ... ]

Seems like an identified can't really be tied to a base domain, or changes to that domain (like proxy or environments) will be really tough. Will we save a non URI based identifier that is more portable that can be used to generate `@id` identifiers using some configuration?

In general though, problems start to arise when you can't rely on the router in order to dereference an `@id`, so for any embedded documents or adding relations and on the client side they may have to know how to construct a URI from an `@id` they are given.

If you do store an identifier that is not tied to a domain/protocol/host, I would suggest that you also include it in the body of the document under something like `dcterms:identifier` or similar.

Reply:
> Should probably call this add, since it's not (visibly) assigning the object to something.

`add` and `remove` are collection (Repository pattern) terms, they make sure an object is in the collection or that it isn't with the intent that only the collection is persisted while objects are transient. `add` then doesn't cover update scenarios because the object is already in the collection, and it is directly updated after a `find`.

A Unit Of Work separately keeps track of which in-memory objects have been modified and should be reflected into persistence. This works seamlessly also for this in-memory implementation, as you retrieve the object and modify it while the map inside here still has a reference to it.

Reply:
We'll need to generate a local UUID or something for each article, but externally the identifier is the URI itself (so isn't something the user will construct). We will probably need a way to be able to find an article by a user-defined identifier though (eg DOI).

Wonder how feasible it is to internally use a different format, then convert it to the public URI on the way out. (Like locally just referencing `/articles/12345`, then making it `http://example.com/prefix/articles/12345`.)

Reply:
The `Repositories` chapter of https://www.oreilly.com/library/view/implementing-domain-driven-design/9780133039900/ talks about Persistence-oriented Repositories (as opposed to Collection-oriented) and in that case a universal (upsert) `save()` method is used along with a `remove()` one.

Reply:
Going to delay the ID question as we can just use blank nodes for now.